File: sample-unity-js-interop-ts/src/tests/test-instance-methods.ts
import { UnityInstance } from "../UnityJSInterop";
import { SampleValues } from "../test-utilities/sample-values";
import { assertArrayEqual, assertEqual } from "../test-utilities/test-asserts";

export function RunTestInstanceMethods(unityInstance: UnityInstance){
    const instance = unityInstance.Module.static.Nahoum.UnityJSInterop.Tests.TestInstanceMethods.GetNewInstance();

    // Get tests
    assertEqual(instance.TestGetDouble().value, SampleValues.TestDouble, "TestInstanceMethods.TestGetDouble");
    assertEqual(instance.TestGetBoolFalse().value, SampleValues.TestBoolFalse, "TestInstanceMethods.TestGetBoolFalse");
    assertEqual(instance.TestGetBoolTrue().value, SampleValues.TestBoolTrue, "TestInstanceMethods.TestGetBoolTrue");
    assertEqual(instance.TestGetInt().value, SampleValues.TestInt, "TestInstanceMethods.TestGetInt");
    assertEqual(instance.TestGetString().value, SampleValues.TestString, "TestInstanceMethods.TestGetString");
    assertEqual(instance.TestGetNullString().value, null, "TestInstanceMethods.TestGetNullString");
    assertEqual(instance.TestGetFloat().value, SampleValues.TestFloat, "TestInstanceMethods.TestGetFloat");
    assertArrayEqual(instance.TestGetFloatArray().value, SampleValues.TestFloatArray, "TestInstanceMethods.TestGetFloatArray");

    // Addition double
    var doubleA = unityInstance.Module.utilities.GetManagedDouble(42);
    var doubleB = unityInstance.Module.utilities.GetManagedDouble(21);
    assertEqual(instance.TestAdditionDouble(doubleA, doubleB).value, 63, "TestInstanceMethods.TestAdditionDouble");

    // Addition float
    var floatA = unityInstance.Module.utilities.GetManagedFloat(42);
    var floatB = unityInstance.Module.utilities.GetManagedFloat(21);
    assertEqual(instance.TestAdditionFloat(floatA, floatB).value, 63, "TestInstanceMethods.TestAdditionFloat");

    // Addition string
    var stringA = unityInstance.Module.utilities.GetManagedString("Hello");
    var stringB = unityInstance.Module.utilities.GetManagedString(" World");
    assertEqual(instance.TestAdditionString(stringA, stringB).value, "Hello World", "TestInstanceMethods.TestAdditionString");
    
}

File: devenv~/Assets/Tests/TestInstanceMethods.cs
using System.Collections.Generic;
using System.Linq;

namespace Nahoum.UnityJSInterop.Tests
{

    public class TestInstanceMethods
    {
        [ExposeWeb] public static TestInstanceMethods GetNewInstance() => new TestInstanceMethods();
        [ExposeWeb] public string TestGetString() => SampleValues.TestString;
        [ExposeWeb] public double TestGetDouble() => SampleValues.TestDouble;
        [ExposeWeb] public int TestGetInt() => SampleValues.TestInt;
        [ExposeWeb] public float TestGetFloat() => SampleValues.TestFloat;
        [ExposeWeb] public float[] TestGetFloatArray() => SampleValues.TestFloatArray;
        [ExposeWeb] public List<float> TestGetFloatList() => SampleValues.TestFloatArray.ToList();
        [ExposeWeb] public double[] TestGetDoubleArray() => SampleValues.TestDoubleArray;
        [ExposeWeb] public bool TestGetBoolTrue() => SampleValues.TestBoolTrue;
        [ExposeWeb] public bool TestGetBoolFalse() => SampleValues.TestBoolFalse;
        [ExposeWeb] public string TestGetNullString() => null;
        [ExposeWeb] public int TestAdditionInt(int a, int b) => (a + b);
        [ExposeWeb] public float TestAdditionFloat(float a, float b) => (a + b);
        [ExposeWeb] public double TestAdditionDouble(double a, double b) => (a + b);
        [ExposeWeb] public string TestAdditionString(string a, string b) => (a + b);

        // Test simple unity types
        [ExposeWeb] public UnityEngine.Vector2 TestGetVector2() => SampleValues.TestVector2;
        [ExposeWeb] public UnityEngine.Vector3 TestGetVector3() => SampleValues.TestVector3;
        [ExposeWeb] public UnityEngine.Vector4 TestGetVector4() => SampleValues.TestVector4;
        [ExposeWeb] public UnityEngine.Quaternion TestGetQuaternion() => SampleValues.TestQuaternion;
        [ExposeWeb] public UnityEngine.Color TestGetColor() => SampleValues.TestColor;
        [ExposeWeb] public UnityEngine.Color32 TestGetColor32() => SampleValues.TestColor32;
    }
}